runtime/tr.source/trj9/z/codegen/J9S390CHelperLinkage.hpp

```diff
@@ -122,50 +122,6 @@ class S390CHelperLinkage : public TR::Linkage
       return buildDirectDispatch(callNode, NULL, returnReg);
       }
 
-   /** \brief
-    *       Overloaded function that builds the direct dispatch sequence for given node using C Helper linkage
-    *       using already evaluated parameters from paramInRegister and is passed stores result in returnReg
-    *
-    *  \param callNode
-    *       TR::Node* for which helper dispatch sequence will be generated
-    *
-    *  \param &paramInRegister
-    *       A stack of TR::Register* which will be used to prepare arguments for helper call instead of evaluating children of callNode
-    * 
-    *  \param returnReg
-    *       TR::Register* allocated by consumer of this API. If passed dispatch sequence will use this register to store return value from helper
-    *       instead from allocating new register.
-    *
-    *  \return
-    *      Returns TR::Register* which contains return value from helper call
-    */  
-   TR::Register* buildDirectDispatch(TR::Node *callNode, TR_Stack<TR::Register*>& paramInRegister, TR::Register *returnReg=NULL)
-   	{
-   	return buildDirectDispatch(callNode, NULL, paramInRegister, returnReg);
-   	}
-
-   /** \brief
-    *       Overloaded function to be used within internal control flow to build the direct dispatch sequence for given node using C Helper linkage
-    *
-    *  \param callNode
-    *       TR::Node* for which helper dispatch sequence will be generated
-    *
-    *  \param **deps
-    *       A register dependency condition which will be filled with helper call condition so that consumer can combine with the dependency condition
-    *       of internal control flow and attach it to merge label
-    * 
-    *  \param returnReg
-    *       TR::Register* allocated by consumer of this API. If passed dispatch sequence will use this register to store return value from helper
-    *       instead from allocating new register.
-    *
-    *  \return
-    *      Returns TR::Register* which contains return value from helper call
-    */  
-   TR::Register *buildDirectDispatch(TR::Node *callNode, TR::RegisterDependencyConditions** deps, TR::Register *returnReg=NULL)
-   	{
-   	TR_Stack<TR::Register*> paramInRegisters(cg()->trMemory());
-   	return buildDirectDispatch(callNode, deps, paramInRegisters, returnReg);
-   	}
-   TR::Register *buildDirectDispatch(TR::Node *callNode, TR::RegisterDependencyConditions** deps, TR_Stack<TR::Register*>& paramInRegisters, TR::Register *returnReg);
+   TR::Register *buildDirectDispatch(TR::Node *callNode, TR::RegisterDependencyConditions** deps, TR::Register *returnReg=NULL);
```

Comment:
Do you think reordering `deps` and `returnReg` would be more intuitive given that we have an API above defined as `buildDirectDispatch(TR::Node *callNode, TR::Register *returnReg)`? The order of parameters may then be consistent.

Reply:
Thanks for pointing this out, It does look more intuitive the way you said.  Let me change it. 
Currently we are encountering following uses of the interface. 
```
1. buildDirectDispatch(node) -> Standard API
2. buildDirectDispatch(node, register) -> If result register is already allocated
3. buildDirectDispatch(node, &deps, register=NULL) -> It is called within ICF and can be used same if we have already allocated register)
```

Reply:
Ok, So checked how it looks if we change the order
```
TR::Register* buildDirectDispatch(TR::Node *node, TR::Register *register, TR::RegisterDependencyConditions ** deps = static_cast<TR::RegisterDependencyConditions**>(NULL));

TR::Register * buildDirectDispatch(TR::Node *node)
{
return buildDirectDispatch(node, NULL, static_cast<TR::RegisterDependencyConditions**>(NULL));
}
TR::Register *buildDirectDispatch(TR::Node *node, TR::RegisterDependencyConditions **deps)
{
return buildDirectDispatch(node, NULL, deps);
}
``` 
Even in this way we are ending up with 3 overridden functions. I doubt if this makes that more difference.

Reply:
@r30shah I understand we cannot get away without three overridden functions. All I was pointing towards was that if we have possible permutation of three parameters `A`, `B`, and `C` defining three APIs as:

```
foo(A);
foo(A,B);
foo(A,B,C);
```

May be more intuitive for the user looking for the appropriate API to use than what we have now, which is:

```
foo(A);
foo(A,B);
foo(A,C,B);
```

i.e. `B` is not consistently declared as the second parameter in two of the APIs. Defining a consistent parameter order may help the user to look for a specific API as the APIs evolve and grow. Take a look at `generateS390MemoryReference` as an example here:

https://github.com/eclipse/omr/blob/master/compiler/z/codegen/OMRMemoryReference.hpp#L461-L474

The parameter ordering is not consistent at all. The `cg` parameter sometimes comes first, sometimes second, sometimes last, etc and there are 13 such overloads. Do we believe a user may have an easier time finding the right API to use if all the parameters were consistently ordered?

Reply:
@fjeremic I got your point. And what we have `generateS390MemoryReference` is really inconsistent and I agree with the point that user would have difficult time to find correct signature to call. 
I would say this case is little different as neither `TR::Register*` nor `TR::RegisterDependencyConditons**` is consistently second parameter. 
I want to be able to call 
`buildDirectDispatch(node)`
`buildDirectDispatch(node, &deps)`
`buildDirectDispatch(node, register)`
`buildDirectDispatch(node, &deps, register)`
And the way I achieve this is by declaring function like,
`TR::Register * buildDirectDispatch(TR::Node *node, TR::RegisterDependencyConditions **deps, TR::Register *reg=NULL`
This was I can call the same function `buildDirectDispatch(node, &deps)` with reg defined `NULL`.
And Declared `TR::Register *buildDirectDispatch(TR::Node *node, TR::Register *reg)` is used to call dispatch function without dependency conditions.

Reply:
I see, so there are four permutations. I have overlooked the fact that in `buildDirectDispatch(node, register)` the `register` parameter does not have a `NULL` default. Thanks for the explanation @r30shah!